### PR TITLE
Add lint to check that a CRL does not contain an empty revokedCertificates element

### PR DESCRIPTION
**Pull request by defacto64**
RFC5280 at §5.1.2.6 (Revoked Certificates) requires that:

`When there are no revoked certificates, the revoked certificates list MUST be absent`

This lint checks that the above requirement is satisfied.

### CODE REVIEW

**Comment by aarongable** [comment]
It doesn't appear to me that `asn1.Unmarshal` has a [documented guarantee](https://pkg.go.dev/encoding/asn1#Unmarshal) that a nil slice means "this field was missing" and a present-but-zero-length slice means "this field was present but empty". Therefore I don't think that this asn1-based approach is guaranteed to remain correct, and that using a cryptobyte-based approach would be both simpler (no need to replicate all of the structs) and more obviously correct.

**Comment by defacto64** [comment]
@aarongable , my lint is not based on the assumption that a nil slice means "this field was missing" as you say; on the contrary, it is based on the assumption that a non-nil slice means "this field was found" (although it may be empty). I am achieving this by unmarshalling the DER based on an _ad hoc_ struct. In my opinion this approach is sufficiently sound, and very readable.

At any rate, the `asn1.Unmarshal` function is used _dozens of times_ within the Zlint source code, overall; if it were affected by the indeterminacy that you suggest, it would be quite worrying because it would potentially impact many other pieces of Zlint, in addition to my lint, as well as other important softwares that make use of that function; so I'm somewhat perplexed by what you're saying, and I'd like to hear the opinion of the authors of the encoding/asn1 package except that I don't know who they are and how to contact them....

File: v3/lints/cabf_br/crl_empty_revoked_certificates_ko.pem [hunk 1]
```diff
@@ -0,0 +1,13 @@
+-----BEGIN X509 CRL-----
```
Comment (christopher-henderson) [inline v3/lints/cabf_br/crl_empty_revoked_certificates_ko.pem:1]:
```
Certificate Revocation List (CRL):
        Version 2 (0x1)
        Signature Algorithm: sha256WithRSAEncryption
        Issuer: C = EU, O = Some CA, CN = Fake CA for zlint testing
        Last Update: Apr 12 07:36:43 2024 GMT
        Next Update: Apr 13 07:36:43 2024 GMT
        CRL extensions:
            X509v3 Authority Key Identifier: 
                01:02:03:04:05
            X509v3 CRL Number: 
                291
No Revoked Certificates.
    Signature Algorithm: sha256WithRSAEncryption
    Signature Value:
        13:45:39:87:ce:01:31:60:72:a0:f2:53:10:cd:81:aa:81:80:
        c1:d2:bf:56:b9:ef:7c:94:f4:a7:35:4b:48:0c:2b:bc:90:5d:
        43:ef:f1:bb:4b:3a:33:43:07:0b:a8:29:64:31:fb:05:e4:22:
        f8:cb:cd:dc:b5:ce:59:64:c8:32:d8:b8:50:bc:2a:67:9b:5b:
        d6:31:5c:09:9f:e8:95:b4:41:15:cc:b0:6e:01:a6:48:fc:5d:
        88:64:a6:0d:43:3f:dd:98:22:d7:ce:5d:56:d5:4f:fc:cb:a9:
        92:8f:7c:09:65:bc:41:a3:2a:20:54:25:e5:3c:94:6c:44:fd:
        39:12:02:9c:8c:a8:4a:f6:4b:18:de:c9:c4:83:b9:62:0e:18:
        ae:22:01:d2:37:97:38:f2:d7:1b:2c:74:52:8b:bc:20:63:51:
        94:2a:07:a2:74:bf:02:d2:7e:e8:d0:2e:e3:e7:ac:a9:0b:82:
        47:1f:8b:7d:34:08:75:1b:73:2a:e2:60:d5:dd:58:e6:6e:0a:
        86:1f:b1:31:af:3d:9f:a7:b5:63:7f:3b:30:a9:77:54:f2:ab:
        92:97:03:6a:8a:b0:79:d6:17:5b:94:a2:bb:70:72:fe:06:14:
        ab:c5:ba:a9:9f:2f:2b:71:d1:8b:5c:5b:6d:01:0a:36:41:82:
        a0:94:36:30
-----BEGIN X509 CRL-----
MIIBsTCBmgIBATANBgkqhkiG9w0BAQsFADBDMQswCQYDVQQGEwJFVTEQMA4GA1UE
ChMHU29tZSBDQTEiMCAGA1UEAxMZRmFrZSBDQSBmb3IgemxpbnQgdGVzdGluZxcN
MjQwNDEyMDczNjQzWhcNMjQwNDEzMDczNjQzWjAAoCEwHzAQBgNVHSMECTAHgAUB
AgMEBTALBgNVHRQEBAICASMwDQYJKoZIhvcNAQELBQADggEBABNFOYfOATFgcqDy
UxDNgaqBgMHSv1a573yU9Kc1S0gMK7yQXUPv8btLOjNDBwuoKWQx+wXkIvjLzdy1
zllkyDLYuFC8KmebW9YxXAmf6JW0QRXMsG4Bpkj8XYhkpg1DP92YItfOXVbVT/zL
qZKPfAllvEGjKiBUJeU8lGxE/TkSApyMqEr2SxjeycSDuWIOGK4iAdI3lzjy1xss
dFKLvCBjUZQqB6J0vwLSfujQLuPnrKkLgkcfi300CHUbcyriYNXdWOZuCoYfsTGv
PZ+ntWN/OzCpd1Tyq5KXA2qKsHnWF1uUortwcv4GFKvFuqmfLytx0YtcW20BCjZB
gqCUNjA=
-----END X509 CRL-----
```

File: v3/lints/cabf_br/crl_empty_revoked_certificates_ok.pem [hunk 2]
```diff
@@ -0,0 +1,13 @@
+-----BEGIN X509 CRL-----
```
Comment (christopher-henderson) [inline v3/lints/cabf_br/crl_empty_revoked_certificates_ok.pem:1]:
```
Certificate Revocation List (CRL):
        Version 2 (0x1)
        Signature Algorithm: sha256WithRSAEncryption
        Issuer: C = EU, O = Some CA, CN = Fake CA for zlint testing
        Last Update: Apr 12 07:28:45 2024 GMT
        Next Update: Apr 13 07:28:45 2024 GMT
        CRL extensions:
            X509v3 Authority Key Identifier: 
                01:02:03:04:05
            X509v3 CRL Number: 
                291
No Revoked Certificates.
    Signature Algorithm: sha256WithRSAEncryption
    Signature Value:
        2d:fa:47:c8:5c:90:57:7f:24:87:2f:bc:a6:85:bb:6d:fd:5f:
        05:77:c1:0d:57:ff:98:22:1e:9f:a2:59:67:ab:47:5b:0e:0f:
        56:c9:8c:78:fc:4d:7d:77:b0:52:41:91:5a:d3:6b:c7:38:fb:
        f9:2a:14:7c:05:95:48:fb:84:e2:21:29:1d:ae:3e:9f:0d:2f:
        f5:39:4d:8f:f5:05:24:71:bb:fb:d5:7a:8f:39:ed:cc:e4:5e:
        2a:7b:b5:b5:c5:e6:c6:9b:07:0d:72:09:f7:35:61:dd:3c:02:
        c5:5b:45:37:9f:ea:00:70:6a:1b:87:76:29:21:57:4b:3d:7d:
        75:af:18:b7:d4:52:ad:25:ce:9e:02:0f:2c:03:82:1a:af:fd:
        c3:76:55:6e:5b:20:57:de:45:f1:ae:c5:08:c5:ad:a7:1f:88:
        ba:f8:b0:2b:4b:45:16:fc:5e:ce:d7:be:b5:37:b7:2d:4f:83:
        60:3b:76:44:c2:93:9a:b1:88:32:7c:ab:fe:37:87:45:8e:ef:
        38:81:f0:0d:78:fa:85:13:d7:3c:be:4a:da:c7:f0:a3:ab:b3:
        a3:b0:54:3e:68:d0:23:d8:87:0d:7a:7c:34:2d:8d:2b:8f:e0:
        58:06:06:65:26:e6:82:5f:33:a8:8c:c2:e8:18:4f:21:5c:35:
        95:a6:61:37
-----BEGIN X509 CRL-----
MIIBrzCBmAIBATANBgkqhkiG9w0BAQsFADBDMQswCQYDVQQGEwJFVTEQMA4GA1UE
ChMHU29tZSBDQTEiMCAGA1UEAxMZRmFrZSBDQSBmb3IgemxpbnQgdGVzdGluZxcN
MjQwNDEyMDcyODQ1WhcNMjQwNDEzMDcyODQ1WqAhMB8wEAYDVR0jBAkwB4AFAQID
BAUwCwYDVR0UBAQCAgEjMA0GCSqGSIb3DQEBCwUAA4IBAQAt+kfIXJBXfySHL7ym
hbtt/V8Fd8ENV/+YIh6follnq0dbDg9WyYx4/E19d7BSQZFa02vHOPv5KhR8BZVI
+4TiISkdrj6fDS/1OU2P9QUkcbv71XqPOe3M5F4qe7W1xebGmwcNcgn3NWHdPALF
W0U3n+oAcGobh3YpIVdLPX11rxi31FKtJc6eAg8sA4Iar/3DdlVuWyBX3kXxrsUI
xa2nH4i6+LArS0UW/F7O1761N7ctT4NgO3ZEwpOasYgyfKv+N4dFju84gfANePqF
E9c8vkrax/Cjq7OjsFQ+aNAj2IcNenw0LY0rj+BYBgZlJuaCXzOojMLoGE8hXDWV
pmE3
-----END X509 CRL-----
```

File: v3/testdata/crl_empty_revoked_certificates_ko.pem [hunk 3]
```diff
@@ -0,0 +1,13 @@
+-----BEGIN X509 CRL-----
```
Comment (christopher-henderson) [inline v3/testdata/crl_empty_revoked_certificates_ko.pem:1]:
```
Certificate Revocation List (CRL):
        Version 2 (0x1)
        Signature Algorithm: sha256WithRSAEncryption
        Issuer: C = EU, O = Some CA, CN = Fake CA for zlint testing
        Last Update: Apr 12 07:36:43 2024 GMT
        Next Update: Apr 13 07:36:43 2024 GMT
        CRL extensions:
            X509v3 Authority Key Identifier: 
                01:02:03:04:05
            X509v3 CRL Number: 
                291
No Revoked Certificates.
    Signature Algorithm: sha256WithRSAEncryption
    Signature Value:
        13:45:39:87:ce:01:31:60:72:a0:f2:53:10:cd:81:aa:81:80:
        c1:d2:bf:56:b9:ef:7c:94:f4:a7:35:4b:48:0c:2b:bc:90:5d:
        43:ef:f1:bb:4b:3a:33:43:07:0b:a8:29:64:31:fb:05:e4:22:
        f8:cb:cd:dc:b5:ce:59:64:c8:32:d8:b8:50:bc:2a:67:9b:5b:
        d6:31:5c:09:9f:e8:95:b4:41:15:cc:b0:6e:01:a6:48:fc:5d:
        88:64:a6:0d:43:3f:dd:98:22:d7:ce:5d:56:d5:4f:fc:cb:a9:
        92:8f:7c:09:65:bc:41:a3:2a:20:54:25:e5:3c:94:6c:44:fd:
        39:12:02:9c:8c:a8:4a:f6:4b:18:de:c9:c4:83:b9:62:0e:18:
        ae:22:01:d2:37:97:38:f2:d7:1b:2c:74:52:8b:bc:20:63:51:
        94:2a:07:a2:74:bf:02:d2:7e:e8:d0:2e:e3:e7:ac:a9:0b:82:
        47:1f:8b:7d:34:08:75:1b:73:2a:e2:60:d5:dd:58:e6:6e:0a:
        86:1f:b1:31:af:3d:9f:a7:b5:63:7f:3b:30:a9:77:54:f2:ab:
        92:97:03:6a:8a:b0:79:d6:17:5b:94:a2:bb:70:72:fe:06:14:
        ab:c5:ba:a9:9f:2f:2b:71:d1:8b:5c:5b:6d:01:0a:36:41:82:
        a0:94:36:30
-----BEGIN X509 CRL-----
MIIBsTCBmgIBATANBgkqhkiG9w0BAQsFADBDMQswCQYDVQQGEwJFVTEQMA4GA1UE
ChMHU29tZSBDQTEiMCAGA1UEAxMZRmFrZSBDQSBmb3IgemxpbnQgdGVzdGluZxcN
MjQwNDEyMDczNjQzWhcNMjQwNDEzMDczNjQzWjAAoCEwHzAQBgNVHSMECTAHgAUB
AgMEBTALBgNVHRQEBAICASMwDQYJKoZIhvcNAQELBQADggEBABNFOYfOATFgcqDy
UxDNgaqBgMHSv1a573yU9Kc1S0gMK7yQXUPv8btLOjNDBwuoKWQx+wXkIvjLzdy1
zllkyDLYuFC8KmebW9YxXAmf6JW0QRXMsG4Bpkj8XYhkpg1DP92YItfOXVbVT/zL
qZKPfAllvEGjKiBUJeU8lGxE/TkSApyMqEr2SxjeycSDuWIOGK4iAdI3lzjy1xss
dFKLvCBjUZQqB6J0vwLSfujQLuPnrKkLgkcfi300CHUbcyriYNXdWOZuCoYfsTGv
PZ+ntWN/OzCpd1Tyq5KXA2qKsHnWF1uUortwcv4GFKvFuqmfLytx0YtcW20BCjZB
gqCUNjA=
-----END X509 CRL-----
```

File: v3/testdata/crl_empty_revoked_certificates_ok.pem [hunk 4]
```diff
@@ -0,0 +1,13 @@
+-----BEGIN X509 CRL-----
```
Comment (christopher-henderson) [inline v3/testdata/crl_empty_revoked_certificates_ok.pem:1]:
```
ertificate Revocation List (CRL):
        Version 2 (0x1)
        Signature Algorithm: sha256WithRSAEncryption
        Issuer: C = EU, O = Some CA, CN = Fake CA for zlint testing
        Last Update: Apr 12 07:28:45 2024 GMT
        Next Update: Apr 13 07:28:45 2024 GMT
        CRL extensions:
            X509v3 Authority Key Identifier: 
                01:02:03:04:05
            X509v3 CRL Number: 
                291
No Revoked Certificates.
    Signature Algorithm: sha256WithRSAEncryption
    Signature Value:
        2d:fa:47:c8:5c:90:57:7f:24:87:2f:bc:a6:85:bb:6d:fd:5f:
        05:77:c1:0d:57:ff:98:22:1e:9f:a2:59:67:ab:47:5b:0e:0f:
        56:c9:8c:78:fc:4d:7d:77:b0:52:41:91:5a:d3:6b:c7:38:fb:
        f9:2a:14:7c:05:95:48:fb:84:e2:21:29:1d:ae:3e:9f:0d:2f:
        f5:39:4d:8f:f5:05:24:71:bb:fb:d5:7a:8f:39:ed:cc:e4:5e:
        2a:7b:b5:b5:c5:e6:c6:9b:07:0d:72:09:f7:35:61:dd:3c:02:
        c5:5b:45:37:9f:ea:00:70:6a:1b:87:76:29:21:57:4b:3d:7d:
        75:af:18:b7:d4:52:ad:25:ce:9e:02:0f:2c:03:82:1a:af:fd:
        c3:76:55:6e:5b:20:57:de:45:f1:ae:c5:08:c5:ad:a7:1f:88:
        ba:f8:b0:2b:4b:45:16:fc:5e:ce:d7:be:b5:37:b7:2d:4f:83:
        60:3b:76:44:c2:93:9a:b1:88:32:7c:ab:fe:37:87:45:8e:ef:
        38:81:f0:0d:78:fa:85:13:d7:3c:be:4a:da:c7:f0:a3:ab:b3:
        a3:b0:54:3e:68:d0:23:d8:87:0d:7a:7c:34:2d:8d:2b:8f:e0:
        58:06:06:65:26:e6:82:5f:33:a8:8c:c2:e8:18:4f:21:5c:35:
        95:a6:61:37
-----BEGIN X509 CRL-----
MIIBrzCBmAIBATANBgkqhkiG9w0BAQsFADBDMQswCQYDVQQGEwJFVTEQMA4GA1UE
ChMHU29tZSBDQTEiMCAGA1UEAxMZRmFrZSBDQSBmb3IgemxpbnQgdGVzdGluZxcN
MjQwNDEyMDcyODQ1WhcNMjQwNDEzMDcyODQ1WqAhMB8wEAYDVR0jBAkwB4AFAQID
BAUwCwYDVR0UBAQCAgEjMA0GCSqGSIb3DQEBCwUAA4IBAQAt+kfIXJBXfySHL7ym
hbtt/V8Fd8ENV/+YIh6follnq0dbDg9WyYx4/E19d7BSQZFa02vHOPv5KhR8BZVI
+4TiISkdrj6fDS/1OU2P9QUkcbv71XqPOe3M5F4qe7W1xebGmwcNcgn3NWHdPALF
W0U3n+oAcGobh3YpIVdLPX11rxi31FKtJc6eAg8sA4Iar/3DdlVuWyBX3kXxrsUI
xa2nH4i6+LArS0UW/F7O1761N7ctT4NgO3ZEwpOasYgyfKv+N4dFju84gfANePqF
E9c8vkrax/Cjq7OjsFQ+aNAj2IcNenw0LY0rj+BYBgZlJuaCXzOojMLoGE8hXDWV
pmE3
-----END X509 CRL-----
```

**Review by christopher-henderson — APPROVED** [review]
Thank you for the added effort to define our own deserialization target - it's rarely super easy for ASN1.

Similar to #834 by only nit is adding the human readable text to the test data (which, for this lint, is admittedly not so useful as OpenSSL appears to accept either and simply prints `No Revoked Certificates`).

File: v3/lints/cabf_br/crl_empty_revoked_certificates_ko.pem (hunk 1 above)
Comment (defacto64) [reply v3/lints/cabf_br/crl_empty_revoked_certificates_ko.pem:1]: Sure

File: v3/testdata/crl_empty_revoked_certificates_ko.pem (hunk 3 above)
Comment (defacto64) [reply v3/testdata/crl_empty_revoked_certificates_ko.pem:1]: Yes

File: v3/lints/cabf_br/crl_empty_revoked_certificates_ok.pem (hunk 2 above)
Comment (defacto64) [reply v3/lints/cabf_br/crl_empty_revoked_certificates_ok.pem:1]: Sure

File: v3/testdata/crl_empty_revoked_certificates_ok.pem (hunk 4 above)
Comment (defacto64) [reply v3/testdata/crl_empty_revoked_certificates_ok.pem:30]: Sure

**Comment by defacto64** [comment]
I have added the the human readable text to the test data, but something is causing 3 out of 4 checks to fail... not sure what.

**Comment by aaomidi** [comment]
Error is:
lints/rfc/lint_crl_revoked_certificates_field_empty.go:46:6: NewEmptyRevokedCertificates redeclared in this block	lints/rfc/lint_crl_empty_revoked_certificates.go:70:6: other declaration of NewEmptyRevokedCertificates

**Comment by pgporada** [comment]
> It looks like this was caused by them not using [newLint.sh](https://github.com/zmap/zlint/blob/master/v3/newLint.sh) which would have generated for them a constructor called NewRevokedCertificates.

I apologize, sorry about that. I did not use `newLint.sh`.

**Comment by christopher-henderson** [comment]
No worries @pgporada! It's a helper, not a requirement :smile:

**Comment by defacto64** [comment]
I have changed slightly the name of my constructor, now all checks are passed.

**Review by christopher-henderson — APPROVED** [review]

**Pull request merged**